### PR TITLE
Introduce header validation as concept

### PR DESCRIPTION
**Pull request by edsko**
This then paves the way for introducing additional checks
(which this doesn't do yet).

### CODE REVIEW

**Comment by edsko** [comment]
One annoying thing: EBBs (obviously). Block numbers don't _always_ increase. But if we do this generically, then that generic code must be EBB aware :angry:

**Comment by mrBliss** [comment]
> One annoying thing: EBBs (obviously). Block numbers don't _always_ increase. But if we do this generically, then that generic code must be EBB aware angry

Thinking out loud: what if we make this `blk`-specific using, e.g., `RunNode`? The default is *increasing*, but we can override the implementation for Byron to handle EBBs correctly.

**Comment by edsko** [comment]
No, it doesn't belong in `RunNode`. We need an abstraction that captures this. It's part of the ledger definition.

File: ouroboros-consensus/src/Ouroboros/Consensus/HeaderValidation.hs [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+headerStatePush (SecurityParam k) chainState newTip HeaderState{..} =
+    case trim pushed of
+      Nothing                   -> HeaderState chainState pushed  headerStateAnchor
+      Just (newAnchor, trimmed) -> HeaderState chainState trimmed (At newAnchor)
+  where
+    pushed :: StrictSeq (AnnTip blk)
+    pushed = headerStateTips :|> newTip
+
+    trim :: StrictSeq (AnnTip blk) -> Maybe (AnnTip blk, StrictSeq (AnnTip blk))
+    trim tips = case Seq.splitAtEnd (fromIntegral k) tips of
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/HeaderValidation.hs]: Avoid `splitAtEnd` in favour of a pattern match (assuming at most `k`).

File: ouroboros-consensus/src/Ouroboros/Consensus/HeaderValidation.hs [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+--
+--   For example, for Praos this means checking the VRF proofs.
+--
+-- o It verifies the 'HasHeader' part of the header.
+--
+--   Specifically, we verify that
+--
+--   x Block numbers are consecutive
+--   x The block number of the first block is 'genesisBlockNo'
+--   x Hashes line up
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/HeaderValidation.hs]: Mention slot numbers

File: ouroboros-consensus/src/Ouroboros/Consensus/HeaderValidation.hs [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+-- o It verifies the consensus part of the header.
+--
+--   For example, for Praos this means checking the VRF proofs.
+--
+-- o It verifies the 'HasHeader' part of the header.
+--
+--   Specifically, we verify that
+--
+--   x Block numbers are consecutive
+--   x The block number of the first block is 'genesisBlockNo'
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/HeaderValidation.hs]: Get rid of `genesisBlockNo`

File: ouroboros-consensus/src/Ouroboros/Consensus/HeaderValidation.hs [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+--   x Hashes line up
+--
+--   Note that this check is independent from both the consensus protocol and
+--   from the choice of ledger.
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/HeaderValidation.hs]: Delete this paragraph ^^

File: ouroboros-consensus/src/Ouroboros/Consensus/Block.hs [hunk 5]
```diff
@@ -123,7 +123,11 @@ class ( GetHeader blk
 data IsEBB
   = IsEBB
   | IsNotEBB
-  deriving (Eq, Show, Generic, NoUnexpectedThunks, Serialise)
+  deriving (Eq, Show, Generic, NoUnexpectedThunks)
+
+instance Serialise IsEBB where
+  encode = encode . fromIsEBB
+  decode = toIsEBB <$> decode
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/Block.hs]: Is this used in network?

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Mock/Block/PBFT.hs [hunk 6]
```diff
@@ -119,6 +119,7 @@ instance ( SimpleCrypto c
   validateView _     = pbftValidateRegular () (simplePBftExt . simpleHeaderExt)
   selectView   _ hdr = (blockNo hdr, IsNotEBB)
 
+
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Mock/Block/PBFT.hs]: Stray extra line

File: ouroboros-consensus/src/Ouroboros/Consensus/Util/CBOR.hs [hunk 7]
```diff
@@ -16,24 +17,41 @@ module Ouroboros.Consensus.Util.CBOR (
   , readIncremental
   , readIncrementalOffsets
   , withStreamIncrementalOffsets
+    -- * Encodin containers
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/Util/CBOR.hs]: Typo

File: ouroboros-consensus/src/Ouroboros/Consensus/Util/CBOR.hs [hunk 8]
```diff
@@ -237,11 +257,55 @@ withStreamIncrementalOffsets hasFS@HasFS{..} decoder fp = \k ->
...
+decodeSeq :: CBOR.D.Decoder s a -> CBOR.D.Decoder s (StrictSeq a)
+decodeSeq f = Seq.fromList <$> decodeList f
+
+encodeMaybe :: (a -> CBOR.E.Encoding) -> Maybe a -> CBOR.E.Encoding
+encodeMaybe enc = \case
+    Nothing -> CBOR.E.encodeListLen 0
+    Just x  -> CBOR.E.encodeListLen 1 <> enc x
+
+decodeMaybe :: CBOR.D.Decoder s a -> CBOR.D.Decoder s (Maybe a)
+decodeMaybe enc = do
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/Util/CBOR.hs]: Argument should be called `dec`

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/LgrDB.hs [hunk 9]
```diff
@@ -179,11 +182,13 @@ defaultArgs fp = LgrDbArgs {
       -- Fields without a default
     , lgrNodeConfig       = error "no default for lgrNodeConfig"
     , lgrDecodeLedger     = error "no default for lgrDecodeLedger"
-    , lgrDecodeChainState = error "no default for lgrDecodeChainState"
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/LgrDB.hs]: Update comment also (maybe also for the chain DB?)

File: ouroboros-consensus/test-util/Test/Util/TestBlock.hs [hunk 10]
```diff
@@ -298,6 +299,12 @@ instance UpdateLedger TestBlock where
 
   ledgerTipPoint = lastAppliedPoint
 
+instance HasAnnTip TestBlock where
+  -- Use defaults
+
+instance ValidateEnvelope TestBlock where
+  firstBlockNo _ = Block.BlockNo 1
```
Comment (edsko) [inline ouroboros-consensus/test-util/Test/Util/TestBlock.hs]: Comment why this is necessary (block number is derived from hash and starts at 1)

File: ouroboros-consensus/src/Ouroboros/Consensus/Block.hs (hunk 5 above)
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/Block.hs]: Nope.

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/LgrDB.hs (hunk 9 above)
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/LgrDB.hs]: the chain DB refers to the documentation for the imm/vol/lgr DB.

**Comment by edsko** [comment]
Rebased and restarted, since Hydra seemed stuck.

**Pull request merged**